pkg/controllers/resourceapply/resourceapply.go

```diff
@@ -0,0 +1,227 @@
+	case *corev1.ConfigMap:
+		return ApplyConfigMap(ctx, client, recorder, t)
+	default:
+		return false, fmt.Errorf("unhandled type %T", resource)
+	}
+}
+
+// ApplyConfigMap merges objectmeta, requires data
+func ApplyConfigMap(ctx context.Context, client coreclientv1.Client, recorder record.EventRecorder, requiredOriginal *corev1.ConfigMap) (bool, error) {
+	required := requiredOriginal.DeepCopy()
+	existing := &corev1.ConfigMap{}
+	err := client.Get(ctx, coreclientv1.ObjectKeyFromObject(requiredOriginal), existing)
+	if apierrors.IsNotFound(err) {
+		err := client.Create(ctx, resourcemerge.WithCleanLabelsAndAnnotations(required).(*corev1.ConfigMap))
+		if err != nil {
+			recorder.Event(required, corev1.EventTypeWarning, "Update failed", err.Error())
+			return false, err
+		}
+		recorder.Event(required, corev1.EventTypeNormal, "Updated successfully", "Resource was successfully updated")
+		return true, nil
+	}
+	if err != nil {
+		recorder.Event(required, corev1.EventTypeWarning, "Update failed", err.Error())
+		return false, err
+	}
+
+	modified := resourcemerge.BoolPtr(false)
+	existingCopy := existing.DeepCopy()
+
+	resourcemerge.EnsureObjectMeta(modified, &existingCopy.ObjectMeta, required.ObjectMeta)
+
+	var modifiedKeys []string
+	for existingCopyKey, existingCopyValue := range existingCopy.Data {
+		if requiredValue, ok := required.Data[existingCopyKey]; !ok || (existingCopyValue != requiredValue) {
+			modifiedKeys = append(modifiedKeys, "data."+existingCopyKey)
+		}
+	}
+	for existingCopyKey, existingCopyBinValue := range existingCopy.BinaryData {
+		if requiredBinValue, ok := required.BinaryData[existingCopyKey]; !ok || !bytes.Equal(existingCopyBinValue, requiredBinValue) {
+			modifiedKeys = append(modifiedKeys, "binaryData."+existingCopyKey)
+		}
+	}
+	for requiredKey := range required.Data {
+		if _, ok := existingCopy.Data[requiredKey]; !ok {
+			modifiedKeys = append(modifiedKeys, "data."+requiredKey)
+		}
+	}
+	for requiredBinKey := range required.BinaryData {
+		if _, ok := existingCopy.BinaryData[requiredBinKey]; !ok {
+			modifiedKeys = append(modifiedKeys, "binaryData."+requiredBinKey)
+		}
+	}
+
+	dataSame := len(modifiedKeys) == 0
+	if dataSame && !*modified {
+		return false, nil
+	}
+	existingCopy.Data = required.Data
+	existingCopy.BinaryData = required.BinaryData
+
+	// at this point we know that we're going to perform a write.  We're just trying to get the object correct
+	toWrite := existingCopy // shallow copy so the code reads easier
+
+	err = client.Update(ctx, toWrite)
+	if err != nil {
+		recorder.Event(required, corev1.EventTypeWarning, "Update failed", err.Error())
+		return false, err
+	}
+	recorder.Event(toWrite, corev1.EventTypeNormal, "Updated successfully", "Resource was successfully updated")
+	return true, err
+}
+
+func ApplyDeployment(ctx context.Context, client coreclientv1.Client, recorder record.EventRecorder, requiredOriginal *appsv1.Deployment) (bool, error) {
```

Comment:
Do we need to have this functions public?

Reply:
It's not necessary but I think it logically makes sense to keep these functions public even if we don't call them directly.

Reply:
We can always make them public later without breaking our API surface, what we can't do is make them private, as otherwise people make have imported them and this would be a breaking change. They should be private until explicitly required to be public IMO

Reply:
they are private now :)